source/common/tcp_proxy/hash_policy.cc

```diff
@@ -0,0 +1,54 @@
+#include "common/tcp_proxy/hash_policy.h"
+#include "envoy/common/exception.h"
+
+namespace Envoy {
+namespace TcpProxy {
+
+class SourceIpHashMethod : public HashPolicyImpl::HashMethod {
+public:
+  absl::optional<uint64_t> evaluate(const Network::Address::Instance* downstream_addr,
+                                    const Network::Address::Instance*) const override {
+    if (downstream_addr && downstream_addr->ip() &&
+        !downstream_addr->ip()->addressAsString().empty()) {
+      return HashUtil::xxHash64(downstream_addr->ip()->addressAsString());
+    }
+
+    return absl::nullopt;
+  }
+};
+
+HashPolicyImpl::HashPolicyImpl(
+    absl::Span<const envoy::config::filter::network::tcp_proxy::v2::TcpProxy::HashPolicy* const>
+        hash_policies) {
+  hash_impls_.reserve(hash_policies.size());
+
+  for (auto* hash_policy : hash_policies) {
+    switch (hash_policy->policy_specifier_case()) {
+    case envoy::config::filter::network::tcp_proxy::v2::TcpProxy::HashPolicy::kSourceIp:
+      hash_impls_.emplace_back(new SourceIpHashMethod());
+      break;
+    default:
+      throw EnvoyException(
+          absl::StrCat("Unsupported hash policy ", hash_policy->policy_specifier_case()));
+    }
+  }
+}
+
+absl::optional<uint64_t>
+HashPolicyImpl::generateHash(const Network::Address::Instance* downstream_addr,
+                             const Network::Address::Instance* upstream_addr) const {
+  absl::optional<uint64_t> hash;
+  for (const HashMethodPtr& hash_impl : hash_impls_) {
+    const absl::optional<uint64_t> new_hash = hash_impl->evaluate(downstream_addr, upstream_addr);
+    if (new_hash) {
+      // Rotating the old value prevents duplicate hash rules from cancelling each other out
```

Comment:
I would rather not define our own hash combination logic here and instead just use Absl::Hash. To avoid this for now, would it be better to put a validation constraint on the repeated hash policy to a single element? Do we even needed a repeated policy?

Reply:
Yeah, putting a validation constraint on the hash policy is reasonable currently. 

> Do we even needed a repeated policy?

Is it possible in the future that someone requires to hash with source port or destination port, even though I didn't find such hash in nginx.